mungegithub/mungers/milestone-maintainer.go

```diff
@@ -0,0 +1,461 @@
+	if !ok {
+		return
+	}
+
+	labelName, notifyState := notificationState(obj, comment, m.botName)
+	// Always attempt to ensure the milestone label
+	if len(labelName) > 0 && !ensureLabel(obj, labelName) {
+		return
+	}
+	if notifyState == nil {
+		// No notification change is required
+		return
+	}
+
+	if notifyState.deleteExisting {
+		if err := obj.DeleteComment(comment.Source.(*githubapi.IssueComment)); err != nil {
+			return
+		}
+	}
+
+	notification := c.Notification{
+		Name:      milestoneNotifierName,
+		Arguments: notifyState.description,
+		Context:   notifyState.message,
+	}
+	if err := notification.Post(obj); err != nil {
+		return
+	}
+
+	if labelName != milestoneRemovedLabel {
+		return
+	}
+	obj.ClearMilestone()
+}
+
+// ignoreObject indicates whether the munger should ignore the given object
+func (m *MilestoneMaintainer) ignoreObject(obj *github.MungeObject) bool {
+	// Only target issues
+	if obj.IsPR() {
+		return true
+	}
+
+	// Only target issues with an assigned milestone
+	milestone, ok := obj.ReleaseMilestone()
+	if !ok || len(milestone) == 0 {
+		return true
+	}
+
+	// Only target issues in the active milestone
+	return milestone != m.activeMilestone
+}
+
+// notificationState returns the state required to ensure the munger's
+// notification is kept current.  If a nil return value is returned,
+// no action should be taken.
+func notificationState(obj *github.MungeObject, comment *c.Comment, botName string) (string, *milestoneNotification) {
+	notification := c.ParseNotification(comment)
+
+	kindLabel, priorityLabel, sigLabels, labelErrors := checkLabels(obj.Issue)
+
+	labelsComplete := len(labelErrors) == 0
+	if labelsComplete {
+		return milestoneLabelsCompleteLabel, labelsCompleteState(obj.Issue, notification, kindLabel, priorityLabel, sigLabels)
+	}
+
+	isBlocker := (priorityLabel == priorityCriticalUrgent)
+	var removeIn *time.Duration
+	// Blockers are never removed from the milestone so grace period computation can be skipped
+	if !isBlocker {
+		removeIn = gracePeriodRemaining(obj, botName)
+		if removeIn == nil {
+			return "", nil
+		}
+	}
+	// removeIn is guaranteed to be non-nil for non-blockers
+	if isBlocker || *removeIn >= 0 {
+		var createdAt *time.Time
+		// createdAt should be nil for blockers to avoid repeatedly posting the same warning message
+		if !isBlocker && comment != nil {
+			createdAt = comment.CreatedAt
+		}
+		return milestoneLabelsIncompleteLabel, labelsIncompleteState(obj.Issue, notification, labelErrors, createdAt, removeIn)
+	}
+	return milestoneRemovedLabel, removalState(obj.Issue, notification, labelErrors, *obj.Issue.Milestone.Title)
+}
+
+// latestNotificationComment returns the most recent notification
+// comment posted by the munger.
+//
+// Since the munger is careful to remove existing comments before
+// adding new ones, only a single notification comment should exist.
+func latestNotificationComment(obj *github.MungeObject, botName string) (*c.Comment, bool) {
```

Comment:
If there is another - separate - comment munger, how will this handle comments from that one?  Will they overwrite another another, or leave each other alone?